### ISSUE TITLE
Graph Operators have colors of edges with 1-based indexing

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#604]
I just noticed that 
```python
# Define custom graph
edge_colors = []
for i in range(L):
    edge_colors.append([i, (i+1)%L, 1])
    edge_colors.append([i, (i+2)%L, 2])

# Define the netket graph object
g = nk.graph.Graph(edges=edge_colors)
```

produces a graph whose colors are `[1,2]` instead of what I would expect `[0,1]`. 
I'm not really opinionated on 1-based indexing, but we shall be consistent.

Fixing this would be a breaking change! So let's do it asap

**Comment by gcarleo** [comment]
thanks for pointing this out, I would also add that we don't really need colors anymore?

**Comment by PhilipVinc** [comment]
why not?

**Comment by gcarleo** [comment]
@VolodyaCO

**Comment by VolodyaCO** [comment]
If I understand correctly, @PhilipVinc you want something like a label (integer) encoding of the colours passed by the user? At this moment it seems that

```python
# Define custom graph
edge_colors = []
for i in range(L):
    edge_colors.append([i, (i+1)%L, 'red'])
    edge_colors.append([i, (i+2)%L, 'blue'])

# Define the netket graph object
g = nk.graph.Graph(edges=edge_colors)
```
should work. I.e. The user can give whatever name they want to their colours. You want that by default colours are mapped to integer labels starting at 0, right?

I can do this, and also raise a warning when the user does not provide integer consecutive labels starting at 0, warning that the colours are re-labelled. Sth like:
```zsh
"You are trying to use 'red' and 'blue' as colours. We have relabelled them to 'red'->0, 'blue'->1"
```

Also, I hadn't noticed that you are now using static typing. If I'm not wrong, I think that those types are not enforced when one runs programs, unless one runs it with something like mypy, right? I see that typing checks are being carried out within methods, i.e. 

```python
def foo(bar: int)-> int:
    assert isinstance(bar, int), "Your argument `bar` is not an integer"
    return bar * 2
```
Maybe we can get rid of those checks with something like typeguard? The change would be minimal, and can be implemented to user exposed functions and methods that run only once (so that speed is not affected):

```python
from typeguard import typechecked

@typechecked
def foo(bar: int)-> int:
    return bar * 2
```

**Comment by PhilipVinc** [comment]
Hmm..
It seems that this is my fault: 
```python
    edge_colors.append([i, (i+1)%L, 1])
    edge_colors.append([i, (i+2)%L, 2])
```

the last element of the tuple is the color, but I had not noticed... 
So this is not a bug after all. Maybe we just need to improve the documentation as it tricked me. Thanks anyway @VolodyaCO for chiming in.(still the proposal to have arbitrary colors is nice)

About type guard: We are not really using static typing. I've started adding type hints to increase the clarity of the code and to generate better documentation. Types are mostly not enforced except for a bunch of places, where it's done manually.
Eventually it would be nice to run mypy on netket, but if you attempt to do so there are several hundreds of errors so we would need to slowly work our way through that. I don't have time for it, though any contribution is welcome.

About your proposal to use type check... that might be a good idea. not everywhere, only in a bunch of places, maybe... 
In case you think about contributing it open another issue to discuss that!
